design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,410 @@
+
+#### CustomResourceDefinitions (CRDS)
+
+CRDs are what Rook is already using to extend Kubernetes.  They are a limited extension mechanism that allows the definition of custom types, but lacks the rich features of API aggregation.  For example, validation of a users CRD is only at the schema level and simple property value checks that are available via the [OpenAPI v3 schema](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject).
+Also, there is currently no versioning (conversion) support.
+However, CRDs are being actively supported by the community and more features are being added to CRDs going forward (e.g., a [versioning proposal](https://github.com/kubernetes/features/issues/544)).
+Finally, CRDs are not a breaking change for Rook users since they are already in use today.
+
+### Controller Options
+
+The controllers are the entities that will perform orchestration and deployment tasks to ensure the users desired state is made a reality within their cluster.
+There are a few options for deploying and hosting the controllers that will perform this work, as explained in the following sections.
+
+#### In-proc Controllers
+
+Multiple controllers could be deployed within a single process.
+For instance, Rook could run one controller per domain of expertise, e.g., ceph-controller, minio-controller, etc.
+Controllers would all watch the same custom resource types via a `SharedInformer` and respond to events via a `WorkQueue` for efficiency and reduced burden on core apiserver.
+Even though all controllers are watching, only the applicable controller responds to and handles an event.
+For example, the user runs `kubectl create cluster.yaml` to create a `cluster.rook.io` instance that has Ceph specific properties.
+All controllers will receive the created event via the `SharedInformer`, but only the Ceph controller will queue and handle it.
+We can consider only loading the controllers the user specifically asks for, perhaps via an environment variable in `rook-operator.yaml`.
+
+Note that this architecture can be used with either API aggregation or CRDs.
+
+##### PROS
+
+* Slightly easier developer burden for new storage backends since there is no need to create a new deployment to host their controller.
+* Less resource burden on K8s cluster since watchers/caches are being shared.
+* Easier migration path to API aggregation in the future, if CRDs usage is continued now.
+
+##### CONS
+
+* All controllers must use same base image since they are all running in the same process.
+* If a controller needs to access a backend specific tool then they will have to schedule a job that invokes the appropriate image.  This is similar to exec’ing a new process but at the cluster level.
+  * Note this only applies to the controller, not to the backend's daemons.  Those daemons will be running a backend specific image and can directly `exec` to their tools.
+
```

Comment:
It would be great to find a way to use a single image for *all* operators. It sounds like invoking one-shot tools is the sticking issue here. Why can't the backend-specific operator start the tools needed in a job or init container of the backend-specific image?

Reply:
Backend specific controllers all running in the same process using the same `rook/rook` image *could* call their specific tools via jobs or init containers, but that adds complexity for something that could be very easy (`exec`).  The model around getting the stdout/stderr, error handling, etc. isn't quite clear to me yet for that approach, but I know it is more difficult and potentially more error prone than a straightforward `exec`.

I think what I'm missing is the real need or value behind using a single image for all controllers.  What does the community gain from forcing that burden on each storage backend? Can you help me understand that better? :)

Reply:
It seems odd to me that a user would have to start each backend-specific operator for things to work. So the experience proposed currently is:
```
kubectl create -f cluster-minio-operator.yaml
kubectl create -f cluster-ceph-operator.yaml
kubectl create -f my-cluster.yaml
```

If the operator is not running for a given backend nothing happens. Or if a give `StorageCluster` has multiple backends they operators must all be running. 

I think the following is a better experience:

```
kubectl create -f rook-operator.yaml
# all controllers are available at this point
kubectl create -f my-cluster.yaml
```

I don't think that backend and privilege isolation are strong arguments against the "in-proc controllers" design, especially if the controllers are limited to calling Kubernetes API only. Note that this is similar to how Kubernetes proper packages many different controllers into a controller-manager process.

Finally, I think ceph is a bit unique in that we need to `exec` command line tools to configure it. Other backends like minio might not need that many `exec`s and might have golang client packages that make the job easier. I bet we can find a way to reduce the `exec` for ceph and even come up with a helper library for `exec_in_container` which starts an init container and captures stdout etc.

Reply:
As discussed offline, an operator pod per backend is OK, mirroring the independent CRDs that each backend will have.